src/main/java/com/ibm/watson/modelmesh/ModelMesh.java

```diff
@@ -5541,147 +5488,208 @@ private Runnable rateTrackingTask() {
+                                //System.out.println("DEBUG: rateTrack "+modelId+" threshold="+scaleUpRpms+" current="+rpm);
+                                if (rpm > heavyRpms) ce.setLastHeavyTime(now);
+
+                                ModelRecord mr = registry.get(modelId);
+                                if (mr == null) continue; // not found in registry (not expected)
+                                // note, it's not possible for inst to be in both loaded + failed lists
+                                int loadedCount = mr.getInstanceIds().size();
+                                if (loadedCount == 0) {
+                                    continue; // unexpected - likely mid unload
+                                }
+                                int failedCount = mr.getLoadFailedInstanceIds().size();
+
+                                int candidateInstCount = suitableInstCount - (loadedCount + failedCount);
+                                if (candidateInstCount <= 0) continue; // no space to load new one
+
+                                // For 1->2 copies, scale-up can also be triggered by a pattern of recent usage
+                                // See explanation of CacheEntry#usageSlices
+                                if (loadedCount == 1) {
+                                    // assert mr.getInstanceIds().containsKey(instanceId);
+
+                                    int i1 = ce.earlierUseIteration, i2 = ce.lastUsedIteration;
+                                    // invariants: lower < upper, i1 <= i2
+                                    if (logger.isDebugEnabled()) {
+                                        logger.debug("Second copy trigger evaluation for model " + modelId
+                                            + ": target range [" + lower + ", " + upper + "], I1="
+                                                + i1 + ", I2=" + i2 + ", curIteration=" + iterationCounter);
+                                    }
+
+                                    boolean i1inRange = false, i2inRange = false;
+                                    if (i2 >= lower && i1 <= upper) {
+                                        i1inRange = i1 >= lower;
+                                        i2inRange = i2 <= upper;
+                                    }
+                                    if (i2inRange || !i1inRange) {
+                                        ce.earlierUseIteration = i2;
+                                    }
+                                    ce.lastUsedIteration = iterationCounter;
+
+                                    if (i1inRange || i2inRange) {
+                                        // Model was used within the target range [MIN_AGE, MAX_AGE] iterations ago
+                                        // so trigger loading of a second copy
+
+                                        // Don't do it if > 90% full and cache is younger than 6 hours
```

Comment:
🤔🤔🤔 

I guess 90% seems reasonable, except maybe in the case where models are really small, like <0.1% of the cache each?

Should the 6 hours here be configurable too?

Super unclear where to draw the line between having way too many levers and having a user complain that 6 hours here is not great because they want their global lru to stay in the days

Reply:
> I guess 90% seems reasonable, except maybe in the case where models are really small, like <0.1% of the cache each?

This would still be _relatively_ close to full, and the cutoff would only apply if all of the very large number of models had been used recently (so still kind of reasonable behaviour).

> Super unclear where to draw the line between having way too many levers and having a user complain that 6 hours here is not great because they want their global lru to stay in the days

Exactly, an overarching goal is to avoid/minimize this kind of configuration as much as possible and just have it adapt and work well depending on the various workload parameters in each case.